ip_country/src/bit_queue.rs

```diff
@@ -0,0 +1,413 @@
+use std::collections::VecDeque;
+
+#[derive(Debug)]
+pub struct BitQueue {
+    back_blank_bit_count: usize, // number of high-order bits in the back byte of the queue that are unused
+    byte_queue: VecDeque<u8>,
+    front_blank_bit_count: usize, // number of low-order bits in the front byte of the queue that are unused
+}
+
+impl BitQueue {
+    pub fn new() -> Self {
+        let byte_queue = VecDeque::from(vec![0, 0]);
+        Self {
+            back_blank_bit_count: 8,
+            byte_queue,
+            front_blank_bit_count: 8,
+        }
+    }
+
+    pub fn len(&self) -> usize {
+        (self.byte_queue.len() * 8) - self.back_blank_bit_count - self.front_blank_bit_count
+    }
+
+    #[allow(unused_assignments)]
+    pub fn add_bits(&mut self, mut bit_data: u64, mut bit_count: usize) {
+        if bit_count > 64 {
+            panic!(
+                "You can only add bits up to 64 at a time, not {}",
+                bit_count
+            )
+        }
+        let initial_bits_added = self.add_some_back_bits(bit_data, bit_count);
+        bit_data >>= initial_bits_added;
+        bit_count -= initial_bits_added;
+        let byte_bits_added = self.add_back_bytes(bit_data, bit_count);
+        bit_data >>= byte_bits_added;
+        bit_count -= byte_bits_added;
+        let final_bits_added = self.add_some_back_bits(bit_data, bit_count);
+        bit_data >>= final_bits_added;
+        bit_count -= final_bits_added;
+        if bit_count != 0 {
+            panic!("Didn't add all the bits: {} left", bit_count);
+        }
+    }
+
+    #[allow(unused_assignments)]
+    pub fn take_bits(&mut self, mut bit_count: usize) -> Option<u64> {
+        let original_bit_count = bit_count;
+        if bit_count > 64 {
+            panic!(
+                "You can only take bits up to 64 at a time, not {}",
+                bit_count
+            )
+        }
+        if bit_count > self.len() {
+            return None;
+        }
+        let mut bit_data = 0u64;
+        let mut bit_position = 0usize;
+        let (initial_bit_data, initial_bit_count) = self.take_some_front_bits(bit_count);
+        bit_data |= initial_bit_data << bit_position;
+        bit_position += initial_bit_count;
+        bit_count -= initial_bit_count;
+        let (byte_bit_data, byte_bit_count) = self.take_front_bytes(bit_count);
+        bit_data |= byte_bit_data << bit_position;
+        bit_position += byte_bit_count;
+        bit_count -= byte_bit_count;
```

Comment:
Again, Bert is right. These four cases should be treated the same. Putting in the `if` after the first `take_some_front_bits()` and the `take_front_bytes()` would make all four cases look the same. Removing the `if`s after the second `take_some_front_bits()` and `take_some_back_bits()` would also make all four cases look the same. None of the `if` statements are strictly necessary. In the `(0, 0)` case, the `if` statement will save a little time; in every other case it will add a little time.